apis/v1alpha1/backendpolicy_types.go

```diff
@@ -0,0 +1,173 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+
+// BackendPolicy defines policies associated with backends. For the purpose of
+// this API, a backend is defined as any resource that a route can forward
+// traffic to. A common example of a backend is a Service. Configuration that is
+// implementation specific may be represented with similar implementation
+// specific custom resources.
+type BackendPolicy struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   BackendPolicySpec   `json:"spec,omitempty"`
+	Status BackendPolicyStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// BackendPolicyList contains a list of BackendPolicy
+type BackendPolicyList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []BackendPolicy `json:"items"`
+}
+
+// BackendPolicySpec defines desired policy for a backend.
+type BackendPolicySpec struct {
+	// BackendRefs define which backends this policy should be applied to. If
+	// more than one BackendPolicy targets the same backend, precedence will be
+	// given to the oldest BackendPolicy.
+	//
+	// Support: Core
+	//
+	// +optional
+	BackendRefs []BackendRef `json:"backendRefs"`
+
+	// TLS is the TLS configuration for these backends.
+	//
+	// Support: Extended
+	//
+	// +optional
+	TLS *BackendTLSConfig `json:"tls,omitempty"`
```

Comment:
@robscott TLS configuration is currently supported by the `Gateway` API and [PR 281](https://github.com/kubernetes-sigs/service-apis/pull/281) proposes adding TLS configuration to the `HTTPRoute` API. Can you help me understand why TLS configuration is required in the `BackendPolicy` API too? If it's required, how does `BackendPolicy` work with the TLS configuration of the other two APIs? cc: @hbagdi

Reply:
TLS in Gateway and in #281 is about downstream TLS i.e. client <-> Gateway.
Backend policy is about upstream TLS i.e. Gateway <-> Upstream Service.

Reply:
We likely need a TLS page/section in the docs to describe all the different places it can be configured.

Reply:
Created https://github.com/kubernetes-sigs/service-apis/issues/331 to track this.